File: src/main/java/com/worldscape/voronoi/VoronoiEdge.java
package com.worldscape.voronoi;

import com.worldscape.voronoi.VoronoiCell;
import com.worldscape.voronoi.VoronoiVertex;
import java.util.Objects;

public class VoronoiEdge {
    private static int idCounter = 0;
    private final String id;
    private VoronoiVertex start;
    private VoronoiVertex end;
    private VoronoiCell leftCell;
    private VoronoiCell rightCell;
    private boolean visible = true;

    public VoronoiEdge(VoronoiVertex start, VoronoiVertex end) {
        this.id = "e_" + idCounter++;
        this.start = start;
        this.end = end;
    }

    public VoronoiEdge(String id, VoronoiVertex start, VoronoiVertex end) {
        this.id = id;
        this.start = start;
        this.end = end;
    }

    public String getId() {
        return this.id;
    }

    public VoronoiVertex getStart() {
        return this.start;
    }

    public void setStart(VoronoiVertex start) {
        this.start = start;
    }

    public VoronoiVertex getEnd() {
        return this.end;
    }

    public void setEnd(VoronoiVertex end) {
        this.end = end;
    }

    public VoronoiCell getLeftCell() {
        return this.leftCell;
    }

    public void setLeftCell(VoronoiCell leftCell) {
        this.leftCell = leftCell;
    }

    public VoronoiCell getRightCell() {
        return this.rightCell;
    }

    public void setRightCell(VoronoiCell rightCell) {
        this.rightCell = rightCell;
    }

    public boolean isVisible() {
        return this.visible;
    }

    public void setVisible(boolean visible) {
        this.visible = visible;
    }

    public double getLength() {
        if (this.start == null || this.end == null) {
            return Double.MAX_VALUE;
        }
        return this.start.distanceTo(this.end);
    }

    public double getMidX() {
        if (this.start == null || this.end == null) {
            return 0.0;
        }
        return (this.start.getX() + this.end.getX()) / 2.0;
    }

    public double getMidY() {
        if (this.start == null || this.end == null) {
            return 0.0;
        }
        return (this.start.getY() + this.end.getY()) / 2.0;
    }

    public boolean isInfinite() {
        return this.start == null || this.end == null;
    }

    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof VoronoiEdge)) {
            return false;
        }
        VoronoiEdge edge = (VoronoiEdge)o;
        return Objects.equals(this.start, edge.start) && Objects.equals(this.end, edge.end);
    }

    public int hashCode() {
        return Objects.hash(this.start, this.end);
    }

    public String toString() {
        return "VoronoiEdge{id='" + this.id + "', start=" + (this.start != null ? this.start.getId() : "null") + ", end=" + (this.end != null ? this.end.getId() : "null") + "}";
    }
}

File: src/main/kotlin/com/worldscape/voronoi/VoronoiEdge.kt
package com.worldscape.voronoi

class VoronoiEdge(
    val id: String = "e_" + idCounter++,
    var start: VoronoiVertex?,
    var end: VoronoiVertex?
) {
    // Secondary constructor for Java interop (Java cannot use Kotlin default params)
    // 供 Java 互操作的辅助构造函数
    constructor(s: VoronoiVertex?, e: VoronoiVertex?) : this(start = s, end = e)

    companion object {
        @JvmField var idCounter: Int = 0
    }

    var leftCell: VoronoiCell? = null
    var rightCell: VoronoiCell? = null
    @get:JvmName("isVisible")
    var visible: Boolean = true

    val length: Double
        get() = if (start == null || end == null) Double.MAX_VALUE else start!!.distanceTo(end!!)

    val midX: Double
        get() = if (start == null || end == null) 0.0 else (start!!.x + end!!.x) / 2.0

    val midY: Double
        get() = if (start == null || end == null) 0.0 else (start!!.y + end!!.y) / 2.0

    fun isInfinite(): Boolean = start == null || end == null

    override fun equals(other: Any?): Boolean =
        this === other || (other is VoronoiEdge && start == other.start && end == other.end)

    override fun hashCode(): Int = java.util.Objects.hash(start, end)

    override fun toString(): String =
        "VoronoiEdge{id='$id', start=${start?.id ?: "null"}, end=${end?.id ?: "null"}}"
}